# **Predicting Term Deposits Subscription: A Classification & Segmentation Approach for Bank Telemarketing Campaigns**

# Problem Statement:

A bank conducted a telemarketing campaign for one of its financial products called ‘Term Deposits’ to help foster long-term relationships with existing customers. The dataset contains information about all the customers who were contacted during a particular year to open term deposit accounts with the bank.

What is the term Deposit?

Term deposits also called fixed deposits, are the cash investments made for a specific time period ranging from 1 month to 5 years for predetermined fixed interest rates. The fixed interest rates offered for term deposits are higher than the regular interest rates for savings accounts. The customers receive the total amount (investment plus the interest) at the end of the maturity period. Also, the money can only be withdrawn at the end of the maturity period. Withdrawing money before that will result in an added penalty associated, and the customer will not receive any interest returns.


# Business Objective:

**This project implements the machine learning binary classification to predict if the customer will take/** **subscribe a bank term deposit (variable y) by analyzing various customers dimensions & a Data-Driven Approach to Predict the Success of Bank Telemarketing**


# Import Libraries

In [1]:
import pandas as pd 
import numpy as np  
import sidetable
import sklearn
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns
import kaggle
from pathlib import Path
from glob import glob

# Display Settings

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)
import warnings
warnings.filterwarnings('ignore')

# Data Ingestion & Basic Understanding

In [3]:
path = Path.cwd().parent
path_data = path/'data'/'raw'
files = []
for file in path_data.glob('*'):
    files.append(file)

files

[WindowsPath('e:/Bank-Telemarketing/data/raw/.gitkeep'),
 WindowsPath('e:/Bank-Telemarketing/data/raw/bank-additional-full.csv'),
 WindowsPath('e:/Bank-Telemarketing/data/raw/bank-additional-names.txt'),
 WindowsPath('e:/Bank-Telemarketing/data/raw/bank-full.csv'),
 WindowsPath('e:/Bank-Telemarketing/data/raw/info.txt')]

In [4]:
f = open(path_data/'bank-additional-names.txt','rt')
for line in f.readlines():
    print(line.strip())

Citation Request:
This dataset is publicly available for research. The details are described in [Moro et al., 2014].
Please include this citation if you plan to use this database:

[Moro et al., 2014] S. Moro, P. Cortez and P. Rita. A Data-Driven Approach to Predict the Success of Bank Telemarketing. Decision Support Systems, In press, http://dx.doi.org/10.1016/j.dss.2014.03.001

Available at: [pdf] http://dx.doi.org/10.1016/j.dss.2014.03.001
[bib] http://www3.dsi.uminho.pt/pcortez/bib/2014-dss.txt

1. Title: Bank Marketing (with social/economic context)

2. Sources
Created by: Sérgio Moro (ISCTE-IUL), Paulo Cortez (Univ. Minho) and Paulo Rita (ISCTE-IUL) @ 2014

3. Past Usage:

The full dataset (bank-additional-full.csv) was described and analyzed in:

S. Moro, P. Cortez and P. Rita. A Data-Driven Approach to Predict the Success of Bank Telemarketing. Decision Support Systems (2014), doi:10.1016/j.dss.2014.03.001.

4. Relevant Information:

This dataset is based on "Bank Marketing" UC

In [5]:
df = pd.read_csv(path/'data'/'raw'/'bank-additional-full.csv', sep=';')
df.head(1)

age        job  marital education default housing loan    contact month  \
0   56  housemaid  married  basic.4y      no      no   no  telephone   may   

  day_of_week  duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0         mon       261         1    999         0  nonexistent          1.10   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           93.99         -36.40       4.86      5191.00  no

In [6]:
df.tail(1)

age      job  marital            education default housing loan  \
41187   74  retired  married  professional.course      no     yes   no   

        contact month day_of_week  duration  campaign  pdays  previous  \
41187  cellular   nov         fri       239         3    999         1   

      poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
41187  failure         -1.10           94.77         -50.80       1.03   

       nr.employed   y  
41187      4963.60  no

In [7]:
df.sample()

age          job marital education default housing loan   contact  \
30564   39  blue-collar  single  basic.9y      no      no   no  cellular   

      month day_of_week  duration  campaign  pdays  previous     poutcome  \
30564   may         mon       264         3    999         0  nonexistent   

       emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
30564         -1.80           92.89         -46.20       1.35      5099.10  no

In [8]:
df.shape
print('Number of rows:', df.shape[0])
print('Number of columns:', df.shape[1])

Number of rows: 41188
Number of columns: 21


In [9]:
# Remove whitespace from column names
df.columns = df.columns.str.strip()
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

The data is collected by Banco de Portugal via direct phone call marketing campaigns but duration of call is not included, where 20 columns are recorded for 41188 term deposit market campaign.

**Bank Client Data Description:**

- **age**: Age of the client (numeric)
- **job**: Type of job (categorical: 'admin.', 'blue-collar', 'entrepreneur', 'housemaid', 'management', 'retired', 'self-employed', 'services', 'student', 'technician', 'unemployed', 'unknown')
- **marital**: Marital status (categorical: 'divorced', 'married', 'single', 'unknown')
- **education**: Level of education (categorical: 'basic.4y', 'basic.6y', 'basic.9y', 'high.school', 'illiterate', 'professional.course', 'university.degree', 'unknown')
- **default**: Whether the client has credit in default (binary: 'no', 'yes', 'unknown')
- **housing**: Whether the client has a housing loan (binary: 'no', 'yes', 'unknown')
- **loan**: Whether the client has a personal loan (binary: 'no', 'yes', 'unknown')

**Last Contact Data Description:**

- **contact**: Type of communication used to contact the client (categorical: 'cellular', 'telephone')
- **month**: Last contact month of the year (categorical: 'jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec')
- **day_of_week**: Last contact day of the week (categorical: 'mon', 'tue', 'wed', 'thu', 'fri')
- **duration**: Duration of the last contact in seconds (numeric)
Important note:  this attribute highly affects the output target (e.g., if duration=0 then y="no"). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model

**Campaign Data Description:**

- **campaign**: Number of contacts performed during this campaign for this client (numeric, includes last contact)
- **pdays**: Number of days since the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
- **previous**: Number of contacts performed before this campaign for this client (numeric) 
- **poutcome**: Outcome of the previous marketing campaign (categorical: 'failure', 'nonexistent', 'success')

**Social and Economic Context Data Description:**

- **emp.var.rate**: Employment variation rate (numeric)
- **cons.price.idx**: Consumer price index (numeric)
- **cons.conf.idx**: Consumer confidence index (numeric)
- **euribor3m**: Euribor 3-month rate (numeric)
- **nr.employed**: Number of employees (numeric)

**Target Variable:**
- **y**: Whether the client subscribed to a term deposit (binary: 'no', 'yes')



**This dataset represents the direct marketing campaign(phone calls) of a Portuguese banking institution, where each record represents a client and the outcome of a marketing campaign to convince the client to subscribe a term deposit. The dataset contains both categorical and numerical features describing each client's demographics, behaviroal, social and economic context, with a binary target variable indicating whether the client subscribed to a term deposit. The data spans from May 2008 to November 2010**

**Missing Attribute Values**: There are several missing values in some categorical attributes, all coded with the "unknown" label. These missing values can be treated as a possible class label or using deletion or imputation techniques.


# Formatting and Consistency Checks

In [14]:
pd.DataFrame({'dtype':df.dtypes,
              'nunique': df.nunique(),
              'unique': [df[col].unique() for col in df.columns]})

dtype  nunique  \
age               int64       78   
job              object       12   
marital          object        4   
education        object        8   
default          object        3   
housing          object        3   
loan             object        3   
contact          object        2   
month            object       10   
day_of_week      object        5   
duration          int64     1544   
campaign          int64       42   
pdays             int64       27   
previous          int64        8   
poutcome         object        3   
emp.var.rate    float64       10   
cons.price.idx  float64       26   
cons.conf.idx   float64       26   
euribor3m       float64      316   
nr.employed     float64       11   
y                object        2   

                                                           unique  
age             [56, 57, 37, 40, 45, 59, 41, 24, 25, 29, 35, 5...  
job             [housemaid, services, admin., blue-collar, tec...  
marital                      [married, single, divorced, unknown]  
education       [basic.4y, high.school, basic.6y, basic.9y, pr...  
default                                        [no, unknown, yes]  
housing                                        [no, yes, unknown]  
loan                                           [no, yes, unknown]  
contact                                     [telephone, cellular]  
month           [may, jun, jul, aug, oct, nov, dec, mar, apr, ...  
day_of_week                             [mon, tue, wed, thu, fri]  
duration        [261, 149, 226, 151, 307, 198, 139, 217, 380, ...  
campaign        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 19...  
pdays           [999, 6, 4, 3, 5, 1, 0, 10, 7, 8, 9, 11, 2, 12...  
previous                                 [0, 1, 2, 3, 4, 5, 6, 7]  
poutcome                          [nonexistent, failure, success]  
emp.var.rate    [1.1, 1.4, -0.1, -0.2, -1.8, -2.9, -3.4, -3.0,...  
cons.price.idx  [93.994, 94.465, 93.918, 93.444, 93.798, 93.2,...  
cons.conf.idx   [-36.4, -41.8, -42.7, -36.1, -40.4, -42.0, -45...  
euribor3m       [4.857, 4.856, 4.855, 4.859, 4.86, 4.858, 4.86...  
nr.employed     [5191.0, 5228.1, 5195.8, 5176.3, 5099.1, 5076....  
y                                                       [no, yes]

In [17]:
def highlight_zero_negative(val):
    if val < 0:
        return 'background-color: red; color: white;'  # Negative values styled in red
    elif val == 0:
        return 'background-color: yellow; color: black;'  # Zero values styled in yellow
    else:
        return ''  # No styling for positive values

In [ ]:
# data descripition for numerical columns
df_descb = df.describe(include='number')
df_descb.style.applymap(highlight_zero_negative)

In [21]:
df.describe(include='object')

job  marital          education default housing   loan   contact  \
count    41188    41188              41188   41188   41188  41188     41188   
unique      12        4                  8       3       3      3         2   
top     admin.  married  university.degree      no     yes     no  cellular   
freq     10422    24928              12168   32588   21576  33950     26144   

        month day_of_week     poutcome      y  
count   41188       41188        41188  41188  
unique     10           5            3      2  
top       may         thu  nonexistent     no  
freq    13769        8623        35563  36548

In [27]:
# data duplicates
df.duplicated().sum()

np.int64(12)

In [34]:
df[df.duplicated(keep='last')]#.sort_values(by='age').head(10)

age          job   marital            education  default housing loan  \
1265    39  blue-collar   married             basic.6y       no      no   no   
12260   36      retired   married              unknown       no      no   no   
14155   27   technician    single  professional.course       no      no   no   
16819   47   technician  divorced          high.school       no     yes   no   
18464   32   technician    single  professional.course       no     yes   no   
20072   55     services   married          high.school  unknown      no   no   
20531   41   technician   married  professional.course       no     yes   no   
25183   39       admin.   married    university.degree       no      no   no   
28476   24     services    single          high.school       no     yes   no   
32505   35       admin.   married    university.degree       no     yes   no   
36950   45       admin.   married    university.degree       no      no   no   
38255   71      retired    single    university.degree       no      no   no   

         contact month day_of_week  duration  campaign  pdays  previous  \
1265   telephone   may         thu       124         1    999         0   
12260  telephone   jul         thu        88         1    999         0   
14155   cellular   jul         mon       331         2    999         0   
16819   cellular   jul         thu        43         3    999         0   
18464   cellular   jul         thu       128         1    999         0   
20072   cellular   aug         mon        33         1    999         0   
20531   cellular   aug         tue       127         1    999         0   
25183   cellular   nov         tue       123         2    999         0   
28476   cellular   apr         tue       114         1    999         0   
32505   cellular   may         fri       348         4    999         0   
36950   cellular   jul         thu       252         1    999         0   
38255  telephone   oct         tue       120         1    999         0   

          poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
1265   nonexistent          1.10           93.99         -36.40       4.86   
12260  nonexistent          1.40           93.92         -42.70       4.97   
14155  nonexistent          1.40           93.92         -42.70       4.96   
16819  nonexistent          1.40           93.92         -42.70       4.96   
18464  nonexistent          1.40           93.92         -42.70       4.97   
20072  nonexistent          1.40           93.44         -36.10       4.96   
20531  nonexistent          1.40           93.44         -36.10       4.97   
25183  nonexistent         -0.10           93.20         -42.00       4.15   
28476  nonexistent         -1.80           93.08         -47.10       1.42   
32505  nonexistent         -1.80           92.89         -46.20       1.31   
36950  nonexistent         -2.90           92.47         -33.60       1.07   
38255  nonexistent         -3.40           92.43         -26.90       0.74   

       nr.employed    y  
1265       5191.00   no  
12260      5228.10   no  
14155      5228.10   no  
16819      5228.10   no  
18464      5228.10   no  
20072      5228.10   no  
20531      5228.10   no  
25183      5195.80   no  
28476      5099.10   no  
32505      5099.10   no  
36950      5076.20  yes  
38255      5017.50   no